base/src/net.ext.c

```diff
@@ -630,8 +630,14 @@ void tls_write_cb(uv_write_t *wreq, int status) {
 $R netQ_TLSConnectionD_closeG_local (netQ_TLSConnection self, $Cont c$cont, $action on_close) {
     uv_stream_t *stream = (uv_stream_t *)from$int(self->_stream);
     // fd == -1 means invalid FD and can happen after __resume__
-    if (stream == -1)
+    if (stream == -1) {
+        // The TLS connection may have been closed already when the server
+        // closed the socket. In that case just call the on_close callback.
+        char errmsg[] = "TLS TCP socket already closed?";
+        log_debug(errmsg);
+        on_close->$class->__asyn__(on_close, self);
```

Comment:
We initially thought about raising an exception here instead, but `http.Client` calls `net.TLSConnection.write()` asynchronously. And I guess others would too?

Reply:
I don't think you should blindly accept async write as the correct thing. It's not a complete design as-is...

Reply:
Actually, for close, I'm not sure we even need to close on_close. Like why do we end up here with a -1 stream?

If you call `TLSConnection.close()` 10 times in a row, would we expect the `on_close()` callback to also be called 10 times? I think it makes more sense to treat it as a declarative thing, so if `TLSConnection.close()` is called on a closed connection, the goal of having the connection closed is already reached. We don't actually transition, i.e. we don't go from established -> closed, and so on_close will not be called!?

The stream must have gone to -1 at some earlier point at which the on_error() callback should be called, right?

Reply:
Calling `on_close` even though we're already closed (`stream == -1`) allows for chaining the callbacks in `reconnect()`:

```python
    def reconnect():
        close(_connect)
```